File: packages/android-client/FishjamClient/src/main/java/com/fishjamcloud/client/CommandsQueue.kt
package com.fishjamcloud.client

import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

internal class CommandsQueue {
  private var commandsQueue: ArrayDeque<Command> = ArrayDeque()
  var clientState: ClientState = ClientState.CREATED

  suspend fun addCommand(command: Command) {
    commandsQueue.add(command)
    suspendCoroutine { cont ->
      run {
        command.continuation = cont
        if (commandsQueue.size == 1) {
          command.execute()
        }
      }
    }
  }

  fun finishCommand() {
    val command = commandsQueue.first()
    commandsQueue.removeFirst()
    if (command.clientStateAfterCommand != null) {
      clientState = command.clientStateAfterCommand
    }
    command.continuation?.resume(Unit)
    // TODO: make it iterative, not recursive?
    if (commandsQueue.isNotEmpty()) {
      commandsQueue.first().execute()
    }
  }

  fun finishCommand(commandName: CommandName) {
    if (commandsQueue.isNotEmpty() && commandsQueue.first().commandName == commandName) {
      finishCommand()
    }
  }

  fun finishCommand(commandNames: List<CommandName>) {
    if (commandsQueue.isNotEmpty() && commandNames.contains(commandsQueue.first().commandName)) {
      finishCommand()
    }
  }

  fun onDisconnected() {
    clientState = ClientState.CREATED
    commandsQueue
      .filter { it.isConnectionCommand() }
      .forEach { command -> command.continuation?.resume(Unit) }
    commandsQueue.removeAll { it.isConnectionCommand() }
    if (commandsQueue.isNotEmpty()) {
      commandsQueue.first().execute()
    }
  }

  fun clear() {
    clientState = ClientState.CREATED
    commandsQueue.forEach { command -> command.continuation?.resume(Unit) }
    commandsQueue.clear()
  }
}

File: packages/ios-client/Sources/FishjamClient/CommandsQueue.swift
import Foundation

class CommandsQueue {
    private var commandsQueue = [Command]()
    var clientState = ClientState.created

    func addCommand(_ command: Command) -> DispatchGroup {
        commandsQueue.append(command)
        if commandsQueue.count == 1 {
            command.execute()
        }
        return command.job
    }

    func finishCommand() {
        guard let command = commandsQueue.first else { return }
        let job = command.job
        commandsQueue.removeFirst()
        if let stateAfterCommand = command.clientStateAfterCommand {
            clientState = stateAfterCommand
        }
        job.leave()
        if !commandsQueue.isEmpty {
            commandsQueue.first?.execute()
        }
    }

    func finishCommand(_ commandName: CommandName) {
        if let firstCommand = commandsQueue.first, firstCommand.commandName == commandName {
            finishCommand()
        }
    }

    func finishCommand(_ commandNames: [CommandName]) {
        if let firstCommand = commandsQueue.first, commandNames.contains(firstCommand.commandName) {
            finishCommand()
        }
    }

    func clear(cause: String) {
        commandsQueue.forEach { command in
            command.job.leave()
        }
        commandsQueue.removeAll()
        clientState = .created
    }
}
